waterbox/opera/bizhawk.cpp

```diff
@@ -0,0 +1,402 @@
+		if (cmd == RETRO_ENVIRONMENT_GET_VARIABLE_UPDATE) { return true; }
+		if (cmd == RETRO_ENVIRONMENT_SET_PIXEL_FORMAT) { *((vdlp_pixel_format_e*) data) = VDLP_PIXEL_FORMAT_XRGB8888; return true; }
+		if (cmd == RETRO_ENVIRONMENT_GET_SYSTEM_DIRECTORY) { *((const char**)data) = systemPath; return true; }
+		if (cmd == RETRO_ENVIRONMENT_GET_SAVE_DIRECTORY) { *((const char**)data) = systemPath; return true; }
+		
+		fprintf(stderr, "Unrecognized environment callback command: %u\n", cmd);
+
+		return false;
+}
+
+/// CD Management Logic Start
+#define CDIMAGE_SECTOR_SIZE 2048
+uint32_t _currentSector = 0;
+void (*cd_read_callback)(int32_t lba, void * dest);
+int (*cd_sector_count_callback)();
+ECL_EXPORT void SetCdCallbacks(void (*cdrc)(int32_t lba, void * dest), int (*cdscc)())
+{
+	cd_read_callback = cdrc;
+	cd_sector_count_callback = cdscc;
+}
+
+ECL_EXPORT void ejectCD()
+{
+	xbus_cdrom_plugin(XBP_SET_COMMAND, (void*) CDROM_CMD_EJECT_DISC);
+}
+
+ECL_EXPORT void insertCD()
+{
+	xbus_cdrom_plugin(XBP_SET_COMMAND, (void*) CDROM_CMD_INJECT_DISC);
+	xbus_cdrom_plugin(XBP_SET_COMMAND, (void*) CDROM_CMD_MODE_SET);
+	xbus_cdrom_plugin(XBP_INIT, NULL);
+}
+
+uint32_t cd_get_size(void) {	return cd_sector_count_callback(); }
+void cd_set_sector(const uint32_t sector_) { _currentSector = sector_; }
+void cd_read_sector(void *buf_) {	cd_read_callback(_currentSector, buf_); }
+/// CD Management Logic End
+
+ECL_EXPORT bool Init(const char* gameFilePath, const char* biosFilePath, const char* fontFilePath, int port1Type, int port2Type, int region)
+{ 
+	_gameFilePath = gameFilePath;
+	_biosFilePath = biosFilePath;
+	_fontFilePath = fontFilePath;
+	_port1Type = port1Type;
+	_port2Type = port2Type;
+ _region = region;
+
+	opera_lr_callbacks_set_environment(retro_environment_callback);
+	opera_lr_callbacks_set_input_state(retro_input_state_callback);
+	opera_lr_callbacks_set_input_poll(retro_input_poll_callback);
+	opera_lr_callbacks_set_audio_sample_batch(retro_audio_sample_batch_callback);
+	opera_lr_callbacks_set_video_refresh(retro_video_refresh_callback);
+
+	retro_set_controller_port_device(0, port1Type);
+	retro_set_controller_port_device(1, port2Type);
+	retro_init();
+
+ // Setting cd callbacks
+	opera_cdrom_set_callbacks(cd_get_size, cd_set_sector, cd_read_sector);
+
+	// Loading game file
+	struct retro_game_info game;
+	game.path = _gameFilePath.c_str();
+	auto loadResult = retro_load_game(&game);
+	if (loadResult == false) { fprintf(stderr, "Could not load game: '%s'\n", _gameFilePath.c_str()); return false; }
+
+	//// Getting memory areas
+	/** 2 = wram, 3 = vram*/
+
+	// WRAM
+	_memoryAreas.wram = (uint8_t*)retro_get_memory_data(RETRO_MEMORY_SYSTEM_RAM);
+	_memorySizes.wram = retro_get_memory_size(RETRO_MEMORY_SYSTEM_RAM);
+
+	// VRAM
+	_memoryAreas.vram = (uint8_t*)retro_get_memory_data(RETRO_MEMORY_VIDEO_RAM);
+	_memorySizes.vram = retro_get_memory_size(RETRO_MEMORY_VIDEO_RAM);

```

Comment:
I don't think having these stored in global structs is necessary, just call these functions in `GetMemoryAreas` directly.

Reply:
This is the way I do it across all my projects. It helps me a lot to keep things tidy. I rather keep it consistent on my side